R/check_tbl_spl_n.R

```diff
@@ -0,0 +1,140 @@
+#' for more details.
+#' @export
+check_tbl_spl_n <- function(tbl, round_id, file_path, hub_path) {
+  config_tasks <- hubUtils::read_config(hub_path, "tasks")
+
+  if (isFALSE(has_spls_tbl(tbl)) || isFALSE(hubUtils::is_v3_config(config_tasks))) {
+    return(skip_v3_spl_check(file_path))
+  }
+
+  hash_tbl <- spl_hash_tbl(tbl, round_id, config_tasks)
+  n_ranges <- get_round_spl_n_ranges(config_tasks, round_id)
+
+  n_tbl <- dplyr::group_by(hash_tbl, .data$compound_idx) %>%
+    dplyr::summarise(
+      n = dplyr::n_distinct(.data$output_type_id),
+      mt_id = unique(.data$mt_id)
+    ) %>%
+    dplyr::left_join(n_ranges, by = "mt_id") %>%
+    dplyr::mutate(
+      less = .data$n < .data$n_min,
+      more = .data$n > .data$n_max,
+      out_range = .data$less | .data$more
+    ) %>%
+    dplyr::filter(.data$out_range)
+
+  check <- nrow(n_tbl) == 0L
+
+  if (check) {
+    details <- NULL
+    errors <- NULL
+  } else {
+    errors <- n_mismatch_errors(
+      n_tbl, hash_tbl, tbl
+    )
+    details <- n_mismatch_details(n_tbl)
+  }
+
+  capture_check_cnd(
+    check = check,
+    file_path = file_path,
+    msg_subject = "Required samples per compound idx task",
+    msg_attribute = NULL,
+    msg_verbs = c("present.", "not present."),
+    details = details,
+    errors = errors
+  )
+}
+
+get_round_spl_n_ranges <- function(config_tasks, round_id) {
+  round_mt <- hubUtils::get_round_model_tasks(config_tasks, round_id)
+  purrr::imap(
+    round_mt,
+    ~ {
+      output_type_id_params <- purrr::pluck(
+        .x,
+        "output_type",
+        "sample",
+        "output_type_id_params"
+      )
+
+      if (is.null(output_type_id_params)) {
+        return(NULL)
+      } else {
+        tibble::tibble(
+          mt_id = .y,
+          n_min = output_type_id_params$min_samples_per_task,
+          n_max = output_type_id_params$max_samples_per_task
+        )
+      }
+    }
+  ) %>%
+    purrr::list_rbind()
+}
+
+n_mismatch_errors <- function(n_tbl, hash_tbl, tbl) {
+  tbl <- tbl[tbl$output_type == "sample", names(tbl) != "value"]
+  purrr::map(
+    purrr::set_names(n_tbl$compound_idx),
+    ~ {
+      spl_d <- hash_tbl$output_type_id[hash_tbl$compound_idx == .x][1L]
+      compound_idx_tbl <- tbl[
+        tbl$output_type_id == spl_d,
+        setdiff(names(tbl), c("output_type_id", "value", "output_type"))
+      ]
+      row <- n_tbl[n_tbl$compound_idx == .x, ] %>% as.vector()
+      list(
+        compound_idx = .x,
+        n = row$n,
+        min_samples_per_task = row$n_min,
+        max_samples_per_task = row$n_max,
+        compound_idx_tbl = compound_idx_tbl
+      )
+    }
+  )
+}
+
+
+n_mismatch_details <- function(n_tbl) {
+  cat_msg <- function(compound_idx, type) { # nolint: object_usage_linter
+    switch(type,
+      less = paste0(
+        "File contains less than the minimum required number of samples per task ",
+        "for compound idx{?s} {.val {compound_idx}}"
```

Comment:
As `compound_idx` was change to `output type ids` is the other error messages, will it make sense to do it here too? (if it's not too difficult)

Reply:
Here the check is counting samples per compound_idx so the message is again accurate.

Reply:
Ah yes sorry I was not clear, the check is indeed counting samples per compound_idx so that makes sense. However, the compound_id is not something the user has access to easily. So, I was wondering if it makes sense to simplify which compound id we are talking about (maybe with which columns value it represents, for example: Scenario A, location US, etc.).
Also I realize now that adding the output type id information here might not be the most helpful. Sorry about that

Reply:
Users can access compound_idx using `hubData::expand_model_out_val_grid(..., include_sample_ids = TRUE)` but they also will be when the function that creates a clear table illustrating compound idx structure of a given round/modeling task is available: https://github.com/hubverse-org/hubData/issues/40
I am also going to suggest we move both those functions to `hubValidations` so they are available to users.

More importantly the details of each `compound_idx` structure is available in the `errors` object so users can immediately see what is expected for a sample to count towards a particular `compound_idx`. 

So I would prefer to leave the mention to compound_idx for now as I would only have to re-introduce when https://github.com/hubverse-org/hubData/issues/40 is complete (although when it is complete I will mention it in the documentation)

Reply:
OK sounds good to me, thanks for the additional information!